app/os_android.go

```diff
@@ -1444,3 +1448,51 @@ func Java_org_gioui_Gio_scheduleMainFuncs(env *C.JNIEnv, cls C.jclass) {
 }
 
 func (_ ViewEvent) ImplementsEvent() {}
+
+var foregroundService struct {
+	intent C.jobject
+	mu     sync.Mutex
+	count  int
```

Comment:
This is much better than before, but users can still call their stop functions more than once. I suggest a `map[*int]bool` where each call to `app.Start` registers a `new(int)` token in the map. The stop function then returns eearly if the map no longer contains the token.

Reply:
I'm afraid I don't see how users can call their stop functions more than once as the returned function checks count while ohlding the lock in order to prevent stopService from being called more than once.

There should be only one instance of the GioForegroundService, so multiple calls to Start() only indicate that some part of the application is doing something that needs to continue running, and when all of the associated stop func() have been called should the service be actually stopped.

I don't see why a map is useful here, where a counter suffices.

Reply:
A program may have two unrelated background jobs, say (a) uploading an image and (b) checking for new messages. If they run concurrently, the job count is incremented to 2. A programming mistake causes (a) to calls its stop function twice, decrementing the count to 0, even though (b) is still in progress. The resulting missing notification is very hard to notice.

A map allows us to detect and ignore multiple calls to any particular stop function. In other words, it is no longer a (subtle) mistake to call your stop function twice.

Reply:
ok, this makes sense to me. Thank you for the explanation.

Reply:
(changes made)